src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -270,20 +270,28 @@ void InitializeMethods (Collection<MethodDefinition> methods)
 
 		internal void MarkEntireType (TypeDefinition type, bool includeBaseTypes, in DependencyInfo reason, IMemberDefinition sourceLocationMember)
 		{
+			MarkEntireTypeInternal (type, includeBaseTypes, reason, sourceLocationMember, new HashSet<TypeDefinition> ());
+		}
+
+		private void MarkEntireTypeInternal (TypeDefinition type, bool includeBaseTypes, in DependencyInfo reason, IMemberDefinition sourceLocationMember, HashSet<TypeDefinition> typesAlreadyVisited)
+		{
 #if DEBUG
 			if (!_entireTypeReasons.Contains (reason.Kind))
 				throw new ArgumentOutOfRangeException ($"Internal error: unsupported type dependency {reason.Kind}");
 #endif
 
+			if (!typesAlreadyVisited.Add (type))
```

Comment:
Why not use `Annotations.IsMarked` instead of new temporary cache?

Reply:
The type may as well be already marked before we call `MarkEntireType` - the method must do its job anyway in that case. This is similar to other similar situations - `MarkEntireType` is a helper called when `DynamicallyAccessedMemberTypes.All` is used. If somebody uses just `DynamicallyAccessedMemberTypes.Methods` we also mark all the methods on the type regardless if the type itself has been marked already or not.

Reply:
That's good observation but in that case, we should cache it globally to avoid processing the type multiple times

Reply:
I was going for a minimalistic fix in hope to backporting this to 5.0.
The global cache is probably slightly better, but only slightly - MarkEntireType is used very rarely, so it's not very likely that the cache will have successful hits outside of System.Object.